src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfo.cs

```diff
@@ -209,27 +218,60 @@ private void DetermineNumberHandling(JsonNumberHandling? parentTypeNumberHandlin
                 {
                     JsonNumberHandlingAttribute? attribute = GetAttribute<JsonNumberHandlingAttribute>(MemberInfo);
 
-                    if (attribute != null &&
-                        !ConverterBase.IsInternalConverterForNumberType &&
-                        ((ClassType.Enumerable | ClassType.Dictionary) & ClassType) == 0)
+                    if (attribute != null && !numberHandlingIsApplicable)
                     {
                         ThrowHelper.ThrowInvalidOperationException_NumberHandlingOnPropertyInvalid(this);
                     }
 
                     handling = attribute?.Handling;
                 }
 
-                // Priority 2: Get handling from attribute on parent class type.
-                handling ??= parentTypeNumberHandling;
-
-                // Priority 3: Get handling from JsonSerializerOptions instance.
-                if (!handling.HasValue && Options.NumberHandling != JsonNumberHandling.Strict)
+                if (numberHandlingIsApplicable)
                 {
-                    handling = Options.NumberHandling;
+                    // Priority 2: Get handling from attribute on parent class type.
+                    handling ??= parentTypeNumberHandling;
+
+                    // Priority 3: Get handling from JsonSerializerOptions instance.
+                    if (!handling.HasValue && Options.NumberHandling != JsonNumberHandling.Strict)
+                    {
+                        handling = Options.NumberHandling;
+                    }
+
+                    NumberHandling = handling;
                 }
+            }
+        }
+
+        private bool TypeIsCollectionOfNumbers()
+        {
+            if (((ClassType.Enumerable | ClassType.Dictionary) & ClassType) == 0)
+            {
+                return false;
+            }
 
-                NumberHandling = handling;
+            Type? elementType = ConverterBase.ElementType;
+            Debug.Assert(elementType != null);
+
+            elementType = Nullable.GetUnderlyingType(elementType) ?? elementType;
+
+            if (elementType == typeof(byte) ||
+                elementType == typeof(decimal) ||
+                elementType == typeof(double) ||
+                elementType == typeof(short) ||
+                elementType == typeof(int) ||
+                elementType == typeof(long) ||
+                elementType == typeof(sbyte) ||
+                elementType == typeof(float) ||
+                elementType == typeof(ushort) ||
+                elementType == typeof(uint) ||
+                elementType == typeof(ulong) ||
+                elementType == JsonClassInfo.ObjectType
```

Comment:
Why `ObjectType` here?

Reply:
This is to support object-typed properties/collection elements/types that are number instances at runtime when serializing (supported in the 5.0 codebase).

Reply:
So used with polymorphic support of System.Object that at runtime object.GetType() is a number, like `List<object>`.

It doesn't break for reference types since this code path doesn't execute, however what happens if there is a `List<object>` where the element types are DateTime? Wouldn't this method return true (and perhaps cause issues)?

Reply:
Thanks. It did cause issues due to trying to call `WriteNumberWithCustomHandling()` with these non-number converters, but this is fixed now with this change: https://github.com/dotnet/runtime/pull/41679/commits/2c8adaad6e60ef4dbc0d63f8d3c270d781f4bcef#diff-56b52018b267905bd21daf4c7615b6c7R359. I've added tests for this scenario.